### PR TITLE
travis-ci: deploy RPM / Deb packages

### LINKED ISSUE #117: Deploy PHP7 Deb/RPM packages
Please build the package and place it into tarantool package repository

### PR DESCRIPTION
**Pull request by Totktonada**
Now php-tarantool packages are deployed into two kinds of repositories:
packagecloud.io and to S3 based ones. The download.tarantool.org now
points to the latter, but we keep packagecloud.io for a while to don't
break users, which use it directly (we don't recommend it though). See
[1] for more info about deprecation of our packagecloud.io repositories.

The deployment scripts allows to provide two configurations: production
and staging. The former is for deployment from the master branch and
from a git tag. The latter is for deployments from a developer branch.
It is useful when something should be verified in a specific environment
using a built package or when the deployment process itself was modified
and should be tested.

The difference between production and staging deployment process is how
duplicate package versions are handled. It give an error for production
deployment, but does not for staging. The new package is discarded in
the case for packagecloud.io, but it replaces the old package in S3
repositories.

Read comments in the deployment scripts for more details.

[1]: https://github.com/tarantool/tarantool/issues/4947

Fixes #117

### CODE REVIEW

**Comment by Totktonada** [comment]
Filed several issues re update_repo.sh during working on this PR:

https://github.com/tarantool/tarantool/issues/5096
https://github.com/tarantool/tarantool/issues/5112
https://github.com/tarantool/tarantool/issues/5113
https://github.com/tarantool/tarantool/issues/5114
https://github.com/tarantool/tarantool/issues/5115

**Comment by Totktonada** [comment]
Implemented automatic fetching of a repositories list to push into for both packagecloud and S3 based repositories.

Possible further features:

* Automatically determine list of distributions (parse CI script in tarantool repository?)
* Allow to schedule a periodic job, which will push to new repositories and will build and push for new distributions.

It seems logical to rewrite it from shell to some readable language and move it to packpack tools. I have no time for this now, to be honest.

The initial plan was to create a simple wrapper around `packagecloud` and `update_repo.sh` tools, which will deal with problems we aware of (such as [this one](https://github.com/tarantool/tarantool/issues/3745)), but the attempt to handle everything appears to be not so trivial.

File: .travis/deploy_s3.sh [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+    2>/dev/null | awk '{ print $2; }' || true)"
+# Guess from git repository name
+[ -z "${PRODUCT:-}" ] && PRODUCT="$(git config --get remote.origin.url | \
+    sed -e 's/.*\///' -e 's/.git$//' || true)"
+# Guess from directory name
+[ -z "${PRODUCT:-}" ] && PRODUCT="$(basename "$(pwd)" || true)"
+
+# Add --product option if there is ${PRODUCT} value. Otherwise it
+# will fall back to default 'tarantool'.
+if [ -n "${PRODUCT:-}" ]; then
```
[the remark this answers is not in the document]
Comment (Totktonada) [reply .travis/deploy_s3.sh:186]: It is not necessary here, because the guessing block above will set the variable (possibly to the empty value). However I think it is good style when `set -u` is used. This way this block will work correctly as with the guessing block above as well as without it.

For `DEPLOY_PRODUCTION_S3_ACCESS_KEY_ID` and other variables below: yep, it is necessary, because it may be unset when only staging configuration is provided.

[the remark this answers is not in the document]
Comment (Totktonada) [reply .travis/deploy_s3.sh:186]: I would leave it as is.

**Comment by Totktonada** [comment]
> just interesting why didn't you use the travis-ci deploy internal routine.

There are several reasons, but all of them comes from problems we met with Travis-CI packagecloud deploy script.

1. We should either duplicate deploy rules or use tricky templates / condition language in order to [don't miss](https://github.com/tarantool/tarantool/issues/3745) deployments from tagged revisions. I found all those ways error-prone and prefer to use shell or any other well known language.
2. We don't need to duplicate deploy rules for each target repository (2_2, 2_3, etc). This way is flexible enough to even implement fetching of the repositories list dynamically.
3. packagecloud.io gives 503 http error code from time to time (quite often) when we try to push a package, so some retry strategy should be used to implement reliable deployment process.
4. I want to skip or rewrite duplicated package for staging repository, don't give an error. Otherwise most of CI jobs on developer branches would be red.

To sum up: this way appears to be much more flexible and allows us to solve several existing problems.
